4 add interactive configuration api with author

gschennersie opened the pull request:
Version with @susuhahnml as author.

susuhahnml committed:
initial version developed by susana hahn

susuhahnml committed:
example notebooks

Commit:
Revert "example notebooks"

This reverts commit 3accc9853843c1a0f9ba08c81949bae0867f9092.

susuhahnml committed:
readme

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+Copyright (c) 2022 Siemens AG Oesterreich
+SPDX-License-Identifier: MIT
+
+# Documentation of Interactive configurator
```
```suggestion
# Documentation of Interactive Configurator
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+### Configuration
+
+The Configurations are handled using the `Config` class in [ooasp/config.py](ooasp/config.py). This class provides usefully functionality to load, edit, and visualize Knowledge Bases.
```
```suggestion
The Configurations are handled using the `OOASPConfiguration` class in [ooasp/config.py](ooasp/config.py). This class provides useful functionality to load, edit, and visualize Configurations.
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+### Knowledge Base
+
+The Knowledge Bases are handled using the `KB` class in [ooasp/kb.py](ooasp/kb.py). This class provides usefully functionality to load, edit, and visualize Knowledge Bases. The string representation of the objects of this class are ASP facts.
```
```suggestion
The Knowledge Bases are handled using the `OOASPKnowledgeBase` class in [ooasp/kb.py](ooasp/kb.py). This class provides useful functionality to load, edit, and visualize Knowledge Bases. The string representation of the objects of this class are ASP facts.
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+### Configuration
+
+The Configurations are handled using the `Config` class in [ooasp/config.py](ooasp/config.py). This class provides usefully functionality to load, edit, and visualize Knowledge Bases.
+Furthermore, it can serve as container for partial configurations, complete configurations, configuration checks and available options for a configuration.  The string representation of the objects of this class are ASP facts.
```
```suggestion
Furthermore, it can serve as container for partial configurations, complete configurations, configuration checks and available options for a configuration. The string representation of the objects of this class are ASP facts.
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+### Interactive Configurator
+
+The interactive Configurator API provides all functionality to create a configuration interactively. This is done by constructing an `InteractiveConfigurator` class from [ooasp/interactive.py](ooasp/interactive.py). The process is done in a [multi shot](#multi-shot-approach) way.
```
```suggestion
The interactive Configurator API provides all functionality to create a configuration interactively. This is done by calling the methods provided by the `InteractiveConfigurator` class from [ooasp/interactive.py](ooasp/interactive.py). The process is done in a [multi-shot](#multi-shot-approach) way.
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+1. *Complete:* Getting a complete configuration extending `C`.
+2. *Check:* Checking `C` for errors.
+3. *Options:* Obtaining all possible leafs for the objects, values for the attributes and associations to complete `C`. 
```
```suggestion
3. *Options:* Obtaining all possible classes for the objects, values for the attributes, and associations to complete `C`. 
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+- [ooasp/encodings/ooasp.lp](ooasp/encodings/ooasp.lp)
+
+This file also introduces **external atoms** to differentiate the task that will be performed without the need for re-grounding. The truth value of those atoms will be defined eternally inside the [Interactive Configurator](#interactive-configurator).
```
```suggestion
This file also introduces **external atoms** to differentiate the task that will be performed without the need for re-grounding. The truth value of those atoms will be defined externally inside the [Interactive Configurator](#interactive-configurator).
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+**Externals**
+
+- **`guess`**  The [guessing](#guess) part is active. This external is used instead of the old predicate `ooasp_configure`.
```
```suggestion
- **`guess`**  The [guessing](#guess) part is active. This external is used instead of the old predicate [`ooasp_configure`](../ooasp/ooasp_config.lp).
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+-----|---------
+*Complete*     |   `true`     |
+*Check*    |   `false`     |
+*Options*    |   `true`     |
```
```suggestion
	Value of this external for the different tasks:
	Task |  Value |
	-----|---------
	*Complete*     |   `true`     |
	*Check*    |   `false`     |
	*Options*    |   `true`     |
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+-----|---------
+*Complete*     |   `true`     |
+*Check*    |   `false`     |
+*Options*    |   `true`     |
```
```suggestion
	Value of this external for the different tasks:
	Task |  Value |
	-----|---------
	*Complete*     |   `true`     |
	*Check*    |   `false`     |
	*Options*    |   `true`     |
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+-----|---------
+*Complete*     |   `true`     |
+*Check*    |   `false`     |
+*Options*    |   `false`     |
```
```suggestion
	Value of this external for the different tasks:
	Task |  Value |
	-----|---------
	*Complete*     |   `true`     |
	*Check*    |   `false`     |
	*Options*    |   `false`     |
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+Although this might seam contra-intuitive, this external should be false for the *Checking* task in order to have the `ooasp_cv` atoms as part of a stable model.
+
+- **`check_partial_cv`**  The check part is active for [partial constraints](#partial-constraints). This refers only to the integrity constrains, however `ooasp_cv` atoms are always derived, regardless of the external values.
```
```suggestion
- **`check_partial_cv`**  The check part is active for [partial constraints](#partial-constraints). This refers only to the integrity constraints, however `ooasp_cv` atoms are always derived, regardless of the external values.
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+*Check*    |   `false`     |
+*Options*    |   `true`     |
+
+Although this might seam contra-intuitive, this external should be false for the *Checking* task in order to have the `ooasp_cv` atoms as part of a stable model.
```
```suggestion
Although this might seem contra-intuitive, this external should be false for the *Checking* task in order to have the `ooasp_cv` atoms as part of a stable model.
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+### Guess
+
+The guessing section contains all choices to decide the leaf class of an object, the associations, and the values. Note that the guessing is **grounded incrementally** (See [multi-shot](#multi-shot-approach)).
```
```suggestion
The guessing section contains all choices to decide the class of an object, the associations, and the values. Note that the guessing is **grounded incrementally** (See [multi-shot](#multi-shot-approach)).
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+- [ooasp/encodings/ooasp_check.lp](ooasp/encodings/ooasp_check.lp)
+  
+Constraints are defined using predicate:
```
```suggestion
Constraints are defined using the `ooasp_cv/5` predicate:
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+
+
+**Domain specific constrains** are defined by the user in a different file, which can be provided when the [Interactive Configurator](#interactive-configurator) is created. Note that these will also be **grounded incrementally** (See [multi-shot](#multi-shot-approach)).
```
```suggestion
**Domain specific constraints** are defined by the user in a different file, which can be provided when the [Interactive Configurator](#interactive-configurator) is created. Note that these will also be **grounded incrementally** (see [multi-shot](#multi-shot-approach)).
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+#### Complete constraints
+
+Those constrains that can not longer be corrected by adding new values and associations. These constraints will make a configuration invalid and thus, are not provided as options for the user. By default all constraints are considered of this type, unless stated otherwise (see bellow [Partial constrains](#partial-constraints)).
```
```suggestion
Those constrains that can not longer be corrected by adding new values and associations. These constraints will make a configuration invalid and thus, are not provided as options for the user. By default all constraints are considered of this type, unless stated otherwise (see below [Partial constraints](#partial-constraints)).
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+Those constrains that can not longer be corrected by adding new values and associations. These constraints will make a configuration invalid and thus, are not provided as options for the user. By default all constraints are considered of this type, unless stated otherwise (see bellow [Partial constrains](#partial-constraints)).
+
+*For example: Upper bound constraints or assigning an invalid value*
```
```suggestion
*For example: Upper-bound constraints or assigning an invalid value.*
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+Those constraints that may be violated in the current (partial) configuration but can become satisfied at a later stage of the configuration process. 
+
+*For example: Lower bound constraints, missing value for an attribute.*
```
```suggestion
*For example: Lower-bound constraints, missing value for an attribute.*
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+*For example: Lower bound constraints, missing value for an attribute.*
+
+Partial constraints are ignored when using brave reasoning in order to get all possible options for values and associations. They are only checked when the external `ooasp_partial_cv` is true. Nonetheless, the constraint violation atoms `ooasp_cv` for this constrains will still appear when checking the configuration for errors. 
```
```suggestion
Partial constraints are ignored when using brave reasoning in order to get all possible options for values and associations. They are only checked when the external `ooasp_partial_cv` is true. Nonetheless, the constraint violation atoms `ooasp_cv` for these constraints will still appear when checking the configuration for errors. 
```

rtaupe committed:
chore: add poetry.lock

(should be under version control according to https://python-poetry.org/docs/basic-usage/#commit-your-poetrylock-file-to-version-control)

rtaupe committed:
chore: .gitignore generated by gitignore.io

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+**Externals**
+
+- **`user(ooasp_isa_leaf(CONFIG,LEAFCLASS,ID))`**
+  The user assigned a leaf class to an object `ID`
```
```suggestion
  The user assigned leaf class `LEAFCLASS` to object `ID`
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+  The user associated `ID1` and `ID2`
+
+- **`user(ooasp_attribute_value(CONFIG,N,ID,VALUE))`**
+  The user selected a value for object `ID`
```
```suggestion
  The user selected value `VALUE` for attribute `N` of object `ID`
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+- **`user(ooasp_attribute_value(CONFIG,N,ID,VALUE))`**
+  The user selected a value for object `ID`
+
+The truth value of this externals is defined by the [Interactive Configurator](#interactive-configurator) based on the current (partial) configuration that is being constructed using an object of class [Config](#configuration). 
```
```suggestion
The truth value of these externals is defined by the [Interactive Configurator](#interactive-configurator) based on the current (partial) configuration that is being constructed using an object of class [OOASPConfiguration](#configuration). 
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+----
+
+## Multi shot approach
```
```suggestion
## Multi-shot approach
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+## Multi shot approach
+
+This approach intends to tackle the bottle neck issue of re-grounding rules multiple times after each change. 
```
```suggestion
This approach intends to tackle the bottleneck issue of re-grounding rules multiple times after each change. 
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+
+
+The rules are grounded for every new object that is introduced in an incremental way. All encodings that starting with the directive `#program domain(new_id).` will be grounded when a new object is introduced into the configuration. This new object will have the id `new_id`.
```
```suggestion
The rules are grounded for every new object that is introduced in an incremental way. The subprogram starting with the directive `#program domain(new_id).`, which is split among several encoding files, will be grounded when a new object is introduced into the configuration. This new object will have the id `new_id`.
```

rtaupe commented on asp_interactive_configuration/ooasp/encodings/ooasp_check.lp:
```diff
@@ -0,0 +1,148 @@
+
+%--- (USER INPUT) Attribute value must belong to instance class
+
+ooasp_cv(CONFIG,no_instance_for_attribute,new_id,"Attribute {} not of selected class",(ATTR,)) :- 
+	ooasp_attribute(V,C1,ATTR,T),
+	ooasp_attribute_value(CONFIG,ATTR,new_id,VALUE),
+	not ooasp_isa(CONFIG,C1,new_id).
```
This constraint violation may be wrongly produced if attributes of the same name appear in two different classes.
Attention: This code snippet also appears in `DOC.md`!

rtaupe replied:
issue reported as #8

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+##### Example (Multiple positions)
+
+In this example we want to generate the constraint violations when an association is not of the right class, using on predicate `ooasp_associated(CONFIG,ASSOC,ID1,ID2)`. We must notice that our `new_id` could be either of the two values `ID1` or `ID2`, this means that we need two rules considering both cases.
```
```suggestion
In this example we want to generate the constraint violations when an association is not of the right class, using predicate `ooasp_associated(CONFIG,ASSOC,ID1,ID2)`. We must notice that our `new_id` could be either of the two values `ID1` or `ID2`, this means that we need two rules considering both cases.
```

rtaupe commented on asp_interactive_configuration/ooasp/encodings/ooasp_check.lp:
```diff
@@ -0,0 +1,148 @@
+
+%--- (USER INPUT) Type of association 
+
+ooasp_cv(CONFIG,wrongtypeinassoc,new_id,"Associated by {} but is not of class {}",(ASSOC,C1)) :-
+	ooasp_configuration(V,CONFIG),
+	ooasp_associated(CONFIG,ASSOC,new_id,_),
+	ooasp_assoc(V,ASSOC,C1,C1MIN,C1MAX,C2,C2MIN,C2MAX),
+	not ooasp_isa(CONFIG,C1,new_id).
+
+ooasp_cv(CONFIG,wrongtypeinassoc,new_id,"Associated by {} but is not of class {}",(ASSOC,C2)) :-
+	ooasp_configuration(V,CONFIG),
+	ooasp_associated(CONFIG,ASSOC,_,new_id),
+	ooasp_assoc(V,ASSOC,C1,C1MIN,C1MAX,C2,C2MIN,C2MAX),
+	not ooasp_isa(CONFIG,C2,new_id).
```
I think we have here a similar problem as with the cv `no_instance_for_attribute`. Should we be supporting attributes and association that appear with the same name in different classes?
Attention: This code snippet also appears in `DOC.md`!

rtaupe replied:
issue reported as #8

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+### Accumulative values
+
+There are some rules that depend on atoms that are computed an an accumulative way. One of this values is the **arity**. 
```
```suggestion
There are some rules that depend on atoms that are computed in an accumulative way. One of this values is the **arity**. 
```

rtaupe commented on asp_interactive_configuration/DOC.md:
````diff
@@ -0,0 +1,336 @@
+arity(new_id,ASSOC,ARITY):- ARITY = #count{ID2:ooasp_assoc(V,ASSOC,new_id,_,_,ID2,_,_)}.
+```
+
+This rule will gather the number of objects that `new_id` is associated to in association `ASSOC`. While this is correct at the moment you ground `new_id` this value will not longer be correct when new identifiers are grounded. This is because they are were not part of the aggregate `#count` calculated before. Therefore, associations of `new_id` with any `ID>new_id` will not be counted. 
````
```suggestion
This rule will gather the number of objects that `new_id` is associated to in association `ASSOC`. While this is correct at the moment you ground `new_id` this value will no longer be correct when new identifiers are grounded. This is because they were not part of the aggregate `#count` calculated before. Therefore, associations of `new_id` with any `ID>new_id` will not be counted. 
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+These rules compute the arity as shown before for the two possible positions of the new ID. The first one for position `1` and the second one for position `2`. Unlike in our previous rule, here, we also include the `new_id` as the last argument of `ooasp_arity`. We do so to identify that this value `ARITY` for the arity is only valid when `new_id` was grounded. This value can be seen as a step.
+
+Next, we include any new associations that this `new_id` might have with previously grounded ids and update their corresponding arity. This is done in the following rules:
```
```suggestion
Next, we include any new associations that this `new_id` might have with previously grounded IDs and update their corresponding arity. This is done in the following rules:
```

rtaupe commented on asp_interactive_configuration/DOC.md:
````diff
@@ -0,0 +1,336 @@
+	not ooasp_associated(CONFIG,ASSOC, new_id, ID2).
+```
+
+These rules update the values like a counter without using `#count`. Note that the last argument of the `ooasp_arity` is `new_id`, serving two proposes: first, to identify that this is a valid arity for that moment, and second to ensure that this is a new head that hasn't been grounded before.
````
```suggestion
These rules update the values like a counter without using `#count`. Note that the last argument of the `ooasp_arity` is `new_id`, serving two purposes: first, to identify that this is a valid arity for that moment, and second to ensure that this is a new head that hasn't been grounded before.
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+#### Using accumulative values
+
+If we want to use this accumulative values in another rule there is one more detail to consider. 
```
```suggestion
If we want to use these accumulative values in another rule there is one more detail to consider. 
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+**Example**
+
+Image the rule:
```
```suggestion
Imagine the rule:
```

rtaupe commented on asp_interactive_configuration/DOC.md:
````diff
@@ -0,0 +1,336 @@
+    ooasp_arity(CONFIG,assoc1,1,ID,ARITY,STEP), ARITY!=4.
+```
+
+This rule is saying that for association `assoc1` any the object `ID` in position `1` should have exactly `4` objects associated. We first notice that this is not using the `new_id`. We can include this by doing:
````
```suggestion
This rule is saying that for association `assoc1` the object `ID` in position `1` should have exactly `4` objects associated. We first notice that this is not using the `new_id`. We can include this by doing:
```

rtaupe commented on asp_interactive_configuration/DOC.md:
````diff
@@ -0,0 +1,336 @@
+    ooasp_arity(CONFIG,assoc1,1,new_id,ARITY,new_id), ARITY!=4, active(new_id).
+```
+
+Since we updated the values of `ooasp_arity` of all previous ids as well, we also need to ground the corresponding rules for those, yielding the rule:
````
```suggestion
Since we updated the values of `ooasp_arity` of all previous IDs as well, we also need to ground the corresponding rules for those, yielding the rule:
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+#### Efficiency
+
+Even though the explanation above corresponds to a correct solution, it is not efficient. Generally speaking, using the computed value of an aggregate such as `#count` in the head of the rule accounts for additional grounding. Ideally, this value should be used only in the body of the rule. This observations leads us to a more efficient solution, where the aggregate is used directly in the constr ...
```
```suggestion
Even though the explanation above corresponds to a correct solution, it is not efficient. Generally speaking, using the computed value of an aggregate such as `#count` in the head of the rule accounts for additional grounding. Ideally, this value should be used only in the body of the rule. This observation leads us to a more efficient solution, where the aggregate is used directly in the constraint. 
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+Notice that, as before, this rule must consider the `active(new_id)` to define in which step the constraint is violated. The count, however, does not depend on the `new_id`. 
+
+We use here two axillary predicates to avoid having to repeat the rule:
```
```suggestion
We use here two auxiliary predicates to avoid having to repeat the rule:
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+- **`ooasp_assoc_limit(V,ASSOC,max,OPT,C,CMAX)`**
+  The association `ASSOC` has a maximum limit of `MAX` when an object of class `C` is of class type `OPT`, where `OPT` is `1` or `2` .
+  *Example: `ooasp_assoc_limit(V,ass,min,1,racks,10)`
+	The association `ass` must be from a `racks` element as class `C1` to at least `10` elements*
```
```suggestion
- **`ooasp_assoc_limit(V,ASSOC,max,POS,C,CMAX)`**
  The association `ASSOC` has a maximum limit of `CMAX` for objects of class `C` at `POS`, where `POS` is `1` or `2` and refers to one end of the association (1: left end, 2: right end).
  *Example: `ooasp_assoc_limit(V,ass,min,1,racks,10)`:
	In the association `ass`, each `racks` object must be associated to at least `10` elements*
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+- **`ooasp_associated_general(CONFIG,ASSOC,OPT,ID1,ID2)`**
+  Object `ID1` appearing in position `OPT` of association `ASSOC` is associated to object `ID2`
+  *Example: `ooasp_associated_general(c1,ass,2,23,10)` object `23` is the class `C2` of association `ass` with object `10`*
```
```suggestion
- **`ooasp_associated_general(CONFIG,ASSOC,POS,ID1,ID2)`**
  Object `ID1` appearing in position `POS` of association `ASSOC` is associated to object `ID2`
  *Example: `ooasp_associated_general(c1,ass,2,23,10)`: object `23` of the second class of association `ass` is associated with object `10`*
```

rtaupe commented on asp_interactive_configuration/DOC.md:
```diff
@@ -0,0 +1,336 @@
+
+### Incremental solving 
+
+This multi shot encoding can directly be used for incremental solving. With this idea we can ground for id `1` and try to find a configuration using Task 1 *Complete*, if there is no complete configuration (returns UNSAT) with this many object, we ground for the next object `2` and repeat the process until we have a satisfiable answer. 
```
```suggestion
This multi-shot encoding can directly be used for incremental solving. With this idea we can ground for id `1` and try to find a configuration using Task 1 *Complete*, if there is no complete configuration (returns UNSAT) with this number of objects, we ground for the next object `2` and repeat the process until we have a satisfiable answer. 
```

rtaupe commented on asp_interactive_configuration/README.md:
```diff
@@ -0,0 +1,23 @@
+# Interactive Configuration with ASP
+
+
+Prototype for interactive configuration using ASP initially developed by Susana Hahn (Potassco Solutions) as part of a collaboration between Potassco Solutions and Siemens T.
```
```suggestion
Prototype for interactive configuration using ASP initially developed by Susana Hahn (Potassco Solutions) as part of a collaboration between Potassco Solutions and Siemens.
```

rtaupe commented on asp_interactive_configuration/README.md:
```diff
@@ -0,0 +1,23 @@
+
+The files corresponding to the current version are in [`ooasp`](ooasp).
+
+The package usage and new features are showcased in the jupyter notebooks inside the folder [`usage`](usage)
```
```suggestion
The package usage and new features are showcased in the jupyter notebooks inside the folder [`usage`](usage).
```

rtaupe commented on asp_interactive_configuration/examples/metro/constraints.lp:
```diff
@@ -0,0 +1,167 @@
+% Copyright (c) 2022 Siemens AG Oesterreich
```
I can't seem to add a comment to the `.png` files, so I'm asking here: What is the difference between `metro_v1.png` and `v1.png`? They look rather similar.

rtaupe commented on asp_interactive_configuration/examples/metro/constraints.lp:
```diff
@@ -0,0 +1,167 @@
+    ooasp_isa(C,wagon,W2),
+    W1!=W2.
+
+% standingroom > 0 -> count(handrail)>0
```
According to the name of the following two CVs, `->` should probably be replaced by `<->` or `iff`?

gschennersie replied:
The metro example is still WIP.

rtaupe commented on asp_interactive_configuration/examples/metro/constraints.lp:
```diff
@@ -0,0 +1,167 @@
+
+%*
+deactivated because of performance
+ooas_cv(C,nrpassengers_neq_nrseats_plus_standingroom,W,"{} nr_passengers {} = nr_seats {} = standingroom {}",(new_id,W,NRPASSENGERS,NRSEATS,STANDINGROOM)):-
```
```suggestion
ooasp_cv(C,nrpassengers_neq_nrseats_plus_standingroom,W,"{} nr_passengers {} = nr_seats {} = standingroom {}",(new_id,W,NRPASSENGERS,NRSEATS,STANDINGROOM)):-
```

rtaupe commented on asp_interactive_configuration/examples/metro/constraints.lp:
```diff
@@ -0,0 +1,167 @@
+*%
+
+%
+%* all equal type except special
```
This CV is commented out. What is the difference to the similar constraints above and why was it commented out?

gschennersie replied:
The metro example is work in progress and will be fixed by #9

rtaupe commented on asp_interactive_configuration/examples/metro/constraints.lp:
```diff
@@ -0,0 +1,167 @@
+    ooasp_attribute_value(C,nr_seats,W,NRSEATS),
+    STANDINGROOM = NRPASSENGERS - NRSEATS.
+
+*%
```
Please think again if all these commented-out CVs shall really stay in the file, and if yes, provide a bit of explanation.

rtaupe commented on asp_interactive_configuration/examples/metro/kb.lp:
```diff
@@ -0,0 +1,51 @@
+ooasp_attribute_enum(metro_v1,handrail,handrail_type,standard).
+ooasp_attribute_enum(metro_v1,handrail,handrail_type,premium).
+
+% constraints
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/examples/racks/constraints.lp:
```diff
@@ -0,0 +1,240 @@
+#program domain(new_id).
+
+
+ooasp_cv(C,assoc1_constraint1,new_id,"Asssociated by {} to {} which is not of type {}",(new_id,ASSOC,ID2,C2)) :- 
```
These CV texts do not make much sense to me. Also, there are 4 variables but only 3 placeholders!?

gschennersie replied:
These and similar bugs will be fixed in #10

rtaupe commented on asp_interactive_configuration/examples/racks/constraints.lp:
```diff
@@ -0,0 +1,240 @@
+    P>8.
+
+% moduleII requires moduleV
+% TODO: Is using active(new_id) enough?
```
What exactly does this TODO mean?

gschennersie replied:
If  using active(new_id) is enough for checking the constraint or if there need to be multiple cv rules.

rtaupe commented on asp_interactive_configuration/examples/racks/constraints.lp:
```diff
@@ -0,0 +1,240 @@
+    NROFMODULEII = 0,
+    NROFMODULEV > 0.
+
+% new frame
```
Why are the remaining CVs commented out?

rtaupe commented on asp_interactive_configuration/examples/racks/kb.lp:
```diff
@@ -0,0 +1,89 @@
+%     ooasp_attribute_value(C,frame_position,F1,P),
+%     ooasp_attribute_value(C,frame_position,F2,P).
+
+
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+    @property
+    def domain_size(self)->int:
+        """
+        The domain size, it is computing by couting the nomber of objects in the factbase.
```
```suggestion
        The domain size, it is computed by counting the number of objects in the fact base.
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+        """
+        True if the configuration has any constraint violations, false otherwise
+        """
+        return len(self.constraint_violations) != 0
```
Isn't `self.constraint_violations` a method and should thus be called with parentheses?

gschennersie replied:
This and similar will be checked in #11

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+    def add_domain(self, class_name:str, object_id:int)->Predicate:
+        """
+        Adds a new domain fact `ooasp_domain`
+            Paramters:
```
```suggestion
            Parameters:
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+    def add_leaf(self,object_id:int, class_name:str)->Predicate:
+        """
+        Adds a new leaf predicate to the factbase
+            Paramters:
```
```suggestion
            Parameters:
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+            Returns:
+                The added fact
+            Throws:
+                Exception if the class name is not a leaf class
```
```suggestion
                Exception if the class of the given name is not a leaf class
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+    def add_value(self, object_id:int, attr_name:str, attr_value)->Predicate:
+        """
+        Adds a new attribute value predicate to the factbase
+            Paramters:
```
```suggestion
            Parameters:
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+        Adds a new attribute value predicate to the factbase
+            Paramters:
+                object_id: The identifier for the object
+                attr_name: The name od the attribute
```
```suggestion
                attr_name: The name of the attribute
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+            Paramters:
+                object_id: The identifier for the object
+                attr_name: The name od the attribute
+                attr_value: The value od the attribute
```
```suggestion
                attr_value: The value of the attribute
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+    def remove_leaf(self, object_id)->List[Predicate]:
+        """
+        Removes any leaf predicates from the factbase associated to the object id
+            Paramters:
```
```suggestion
            Parameters:
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+
+    def remove_value(self, object_id:id, attr_name:str)->List[Predicate]:
+        """
+        Removes any exisitng values for the given attribute of an object in the factbase
```
```suggestion
        Removes any existing values for the given attribute of an object in the fact base
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+    def remove_value(self, object_id:id, attr_name:str)->List[Predicate]:
+        """
+        Removes any exisitng values for the given attribute of an object in the factbase
+            Paramters:
```
```suggestion
            Parameters:
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+        Removes any exisitng values for the given attribute of an object in the factbase
+            Paramters:
+                object_id: The identifier for the object
+                attr_name: The name od the attribute
```
```suggestion
                attr_name: The name of the attribute
```

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+
+    def remove_association(self, assoc_name:str, object_id1:int, object_id2:int)->List[Predicate]:
+        """
+        Removes the the association from the factbase
```
```suggestion
        Removes the association from the fact base
```

rtaupe commented on asp_interactive_configuration/ooasp/encodings/ooasp_arity.lp:
```diff
@@ -0,0 +1,46 @@
+
+
+% ooasp_arity(CONFIG, ASSOC, OPT, N, ARITY, STEP)
+% arity of node N for association as class C(OPT) is ARITY in step STEP
```
```suggestion
% ooasp_arity(CONFIG, ASSOC, POS, N, ARITY, STEP)
% arity of association ASSOC for object N of class at position POS is ARITY in step STEP
```

rtaupe commented on asp_interactive_configuration/ooasp/encodings/ooasp_aux_config.lp:
```diff
@@ -0,0 +1,37 @@
+% Auxiliary predicate to avoid duplicated rules
+% ooasp_associated_general(CONFIG,ASSOC,OPT,ID1,ID2)
+%    Object ID1 apearing in position OPT of association ASSOC is associated to object ID2
+%    ooasp_associated_general(c1,ass,2,23,10) object 23 is the class C2 of association ass with object 10
```
```suggestion
% Auxiliary predicate to avoid duplicated rules
% ooasp_associated_general(CONFIG,ASSOC,POS,ID1,ID2)
%    Object ID1 apearing in position POS of association ASSOC is associated to object ID2
%    Example: ooasp_associated_general(c1,ass,2,23,10): : object 23 of the second class of association ass is associated with object 10
```

rtaupe commented on asp_interactive_configuration/ooasp/encodings/ooasp_check.lp:
```diff
@@ -0,0 +1,148 @@
+
+#program domain(new_id).
+
+%------ Errors into constraints
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/encodings/ooasp_check.lp:
```diff
@@ -0,0 +1,148 @@
+	active(new_id).
+
+
+%--- Associations are only allowed to something instanciated
```
```suggestion
%--- Associations are only allowed to something instantiated
```

rtaupe commented on asp_interactive_configuration/ooasp/encodings/ooasp_guess.lp:
```diff
@@ -0,0 +1,47 @@
+	ooasp_configuration(V,CONFIG),
+	guess.
+
+% TODO: this rule is only used if MAX-MIN < 30
```
This seems to be obsolete since line 45 is commented out.

rtaupe commented on asp_interactive_configuration/ooasp/config.py:
```diff
@@ -0,0 +1,386 @@
+from clingraph.graphviz import compute_graphs, render
+from .kb import OOASPKnowledgeBase
+from copy import deepcopy, copy
+import ooasp.utils as utils
```
```suggestion
from typing import List
from types import SimpleNamespace
from clingo import Model, parse_term
from clorm.clingo import Control
from clorm import Symbol, Predicate, ConstantField, IntegerField, FactBase, RawField, refine_field, Raw, StringField
from clingraph.orm import Factbase
from clingraph.graphviz import compute_graphs, render
from .kb import OOASPKnowledgeBase
from copy import deepcopy
import ooasp.utils as utils
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+from typing import List 
+from copy import copy, deepcopy
+import ooasp.utils as utils
+from typing import List
```
```suggestion
import time
from clorm.clingo import Control
from clingo import Number, Function
from clorm import Predicate, unify
from ooasp.config import OOASPConfiguration
from ooasp.kb import OOASPKnowledgeBase
from typing import List 
from copy import deepcopy
import ooasp.utils as utils
from typing import List
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+                kb: The Knowledge base
+                config: The partial or complete configuration
+                domain_size: The number of objects in the partial configuration
+                action: The action perfored to get the state
```
```suggestion
                action: The action performed to get the state
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+        Creates a state by copying the state and changing some properties
+            Parameters:
+                action: The action for the new state
+                deep_config: Wether the configuration should be copied in a deep way
```
```suggestion
                deep_config: Whether the configuration should be copied in a deep way
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+        """
+        Grounds all missing programs based on the current domain_size
+        """
+        # print("Ground missing")
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+        Adds a new fact to the control. The fact is added as part
+        of the program for the current domain size.
+            Parameters:
+                facts: The clorm predicate to be added to the control
```
```suggestion
                fact: The clorm predicate to be added to the control
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+        the given iterarator, otherwise tt will ground any missing steps and solve.
+        Sets the found configuration based on the next computed model.
+            Returns:
+                The found OOASPConfiguration or None if no solution is found
```
```suggestion
        Gets the next available solution for the current configuration without increasing the domain size.
        If there is an ongoing browsing process it will continue using
        the given iterator, otherwise it will ground any missing steps and solve.
        Sets the found configuration based on the next computed model.
            Returns:
                The found OOASPConfiguration or None if no solution is found
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+            self._set_user_externals()
+            self.ctl.configuration.solve.enum_mode = 'auto'
+            start = time.time()
+            # print("\n---Calling handle")
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+        
+        start = time.time()
+        try:
+            # print("---Calling next")
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+        try:
+            # print("---Calling next")
+            model = next(self.solution_iterator)
+            # print("Done")
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+                    self.kb, model)
+            return found_config
+        except StopIteration:
+            # print("Cancel")
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+                representing the possible options
+        """
+        if self.brave_config is None:
+            raise RuntimeError("A brave configuration must be computed to get it as optionss")
```
```suggestion
            raise RuntimeError("A brave configuration must be computed to get it as options")
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+        """
+        self._ground_missing()
+
+        if not self.browsing:
```
Maybe there's something about Python I don't know, but shouldn't the method `self.browsing` be called with parentheses? (Here and in other places)

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+        with  self.ctl.solve(yield_=True) as hdn:
+            for model in hdn:
+                # for s in model.symbols(atoms=True):
+                    # print(s)
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+        Creates a new state.
+        Gets the next avaliable solution for the current configuration without increasing the domain size.
+        If there is an ongoing browsing process it will continue ussing
+        the given iterarator, otherwise tt will ground any missing steps and solve.
```
```suggestion
        If there is an ongoing browsing process it will continue using
        the given iterator, otherwise it will ground any missing steps and solve.
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+            Parameters:
+                object_id: Id of the object
+            Throws:
+                Error in case the leaf_class is not really a leaf
```
```suggestion
                Error in case the leaf_class is not really a leaf class
```

rtaupe commented on asp_interactive_configuration/ooasp/interactive.py:
```diff
@@ -0,0 +1,710 @@
+        The found configuration can be selected as the new one using select_found_configuration
+
+        This will affect the Number of object in the current configuration
+        even if the found condiguration is never selected.
```
```suggestion
        even if the found configuration is never selected.
```

rtaupe commented on asp_interactive_configuration/ooasp/tests/test_config.py:
```diff
@@ -0,0 +1,60 @@
+# Copyright (c) 2022 Siemens AG Oesterreich
+# SPDX-License-Identifier: MIT
+
+from clorm.clingo import Control
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/tests/test_correctness.py:
```diff
@@ -0,0 +1,130 @@
+
+from ooasp.interactive import InteractiveConfigurator
+from ooasp.kb import OOASPKnowledgeBase
+from time import time
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/tests/test_interactive.py:
```diff
@@ -0,0 +1,214 @@
+from clingo import parse_term
+from ooasp.kb import OOASPKnowledgeBase
+from ooasp.config import OOASPConfiguration
+from ooasp.interactive import InteractiveConfigurator
```
```suggestion
import pytest
from clingo import parse_term
from ooasp.kb import OOASPKnowledgeBase
from ooasp.interactive import InteractiveConfigurator
```

rtaupe commented on asp_interactive_configuration/ooasp/tests/test_performance.py:
```diff
@@ -0,0 +1,36 @@
+
+from ooasp.interactive import InteractiveConfigurator
+from ooasp.kb import OOASPKnowledgeBase
+from time import time
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/tests/test_performance.py:
```diff
@@ -0,0 +1,36 @@
+        iconf.new_leaf("elementA")
+    iconf.extend_domain(nr_of_elements + 5)
+    config = iconf.next_solution()
+    # config.save_png()
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/ui.py:
```diff
@@ -0,0 +1,223 @@
+        self.config_image = widgets.Image(format='png')
+        self.found_config = widgets.Image(format='png')
+
+        self.history = VBox(layout=Layout(height='300ox', width='auto',overflow_y='auto'))
```
Is `ox` a typo, should it be `px`?

gschennersie replied:
#11

rtaupe commented on asp_interactive_configuration/ooasp/ui.py:
```diff
@@ -0,0 +1,223 @@
+
+    def update(self):
+        """
+        Uodates the UI based on the current interactive state
```
```suggestion
        Updates the UI based on the current interactive state
```

rtaupe commented on asp_interactive_configuration/ooasp/utils.py:
```diff
@@ -0,0 +1,69 @@
+
+def pretty_dic(d):
+    s = (json.dumps(d, indent=4))
+    # s = ""
+    # for k,v in d.items():
+    #     s += COLORS['GREY'] + k + " : " + str(v) + COLORS['NORMAL'] + "\n"
```
```suggestion
```

rtaupe commented on asp_interactive_configuration/ooasp/ui.py:
```diff
@@ -0,0 +1,223 @@
+# SPDX-License-Identifier: MIT
+
+import ipywidgets as widgets
+from ipywidgets import Button, HBox, VBox, Label, Layout,GridspecLayout, HTML
```
```suggestion
from ipywidgets import Button, VBox, Label, Layout,GridspecLayout, HTML
```

rtaupe commented on asp_interactive_configuration/pyproject.toml:
```diff
@@ -0,0 +1,29 @@
+name = "asp_interactive_configuration"
+version = "0.1.0"
+description = "project to develop an interactive asp-based configuration approach"
+authors = ["gschennersie"]
```
According to https://python-poetry.org/docs/pyproject/, authors must be in the form `name <email>`.

gschennersie replied:
I don't want to provide my email.

rtaupe commented on asp_interactive_configuration/usage/Configuration.ipynb:
```diff
@@ -0,0 +1,252 @@
+     "evalue": "",
+     "output_type": "error",
+     "traceback": [
+      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
```
I suggest to re-run the notebook to see if the error disappears

gschennersie replied:
#11

rtaupe commented on asp_interactive_configuration/usage/KB.ipynb:
```diff
@@ -0,0 +1,380 @@
+     "evalue": "",
+     "output_type": "error",
+     "traceback": [
+      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
```
I suggest to re-run the notebook to see if the error disappears

gschennersie replied:
#11

rtaupe reviewed (changes requested):
@gschennersie, thank you very much for committing this contribution!
Could you please look through my comments before we merge?

rtaupe reviewed (changes requested):
@gschennersie, thank you very much for committing this contribution!
Could you please look through my comments before we merge?

rtaupe commented:
Issues created while reviewing this PR that need not be fixed before the merge:
#6 
#7 
#8

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/utils.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/ui.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie reviewed (commented):
Review finished. Created open issues. Ready to merge.

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/DOC.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/README.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/README.md

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/examples/metro/constraints.lp

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/examples/metro/kb.lp

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/examples/racks/kb.lp

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/encodings/ooasp_arity.lp

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/encodings/ooasp_aux_config.lp

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/encodings/ooasp_check.lp

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/encodings/ooasp_check.lp

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/tests/test_config.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/tests/test_correctness.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/tests/test_interactive.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/tests/test_performance.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/tests/test_performance.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie committed:
Update asp_interactive_configuration/ooasp/ui.py

Co-authored-by: Richard Comploi-Taupe <rtaupe@edu.aau.at>

gschennersie reviewed (commented):
review completed.
